### PR TITLE
fix replace memory op for float80

### PR DESCRIPTION
**Pull request by kumarak**
Fix lowering of memory op intrinsics for float80

Ref: https://github.com/lifting-bits/remill/pull/539

### CODE REVIEW

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: libraries/anvill_passes/src/InstructionFolderPass.cpp [hunk 1]
```diff
@@ -558,6 +557,33 @@ bool InstructionFolderPass::FoldSelectWithCastInst(
...
   CopyMetadataTo(cast_instr, new_false_value);
 
+
+  // For vector select check if the condition element type is correct
+  // and the number of elements in the condition type matches if the
+  // destination type.
+  auto should_fold = [](llvm::Type *condition_type,
+                        llvm::Type *destination_type) {
+    if (auto cond_vt = llvm::dyn_cast<llvm::VectorType>(condition_type)) {
+      if (cond_vt->getElementType() !=
```
[the remark this answers is not in the document]
Comment (kumarak) [reply libraries/anvill_passes/src/InstructionFolderPass.cpp:556]: Returning type checks with `IntegerType` will not cover all cases. There could be a mismatch in number of elements and vector select folding will fail in such cases as well. 

```  
  %181 = select <16 x i1> %164, <16 x i8> zeroinitializer, <16 x i8> %180
  %182 = bitcast <16 x i8> %180 to <2 x i64>
  %183 = bitcast <16 x i8> %181 to <2 x i64>
```

In the above case, there will be mismatch of the destination type elements count causing the folding to fail.

File: libraries/anvill_passes/src/LowerRemillMemoryAccessIntrinsics.cpp [hunk 2]
```diff
@@ -57,6 +57,29 @@ static void ReplaceMemReadOp(llvm::CallBase *call_inst, llvm::Type *val_type) {
...
+  auto ptr = ir.CreateIntToPtr(addr, llvm::PointerType::get(val_type, 0));
+  CopyMetadataTo(call_inst, ptr);
+  llvm::Value *val = ir.CreateLoad(ptr);
+  CopyMetadataTo(call_inst, val);
+
+  auto val_ptr = ir.CreateIntToPtr(call_inst->getArgOperand(2),
+                                   llvm::PointerType::get(val_type, 0));
+  CopyMetadataTo(call_inst, val_ptr);
+  (void) ir.CreateStore(val, val_ptr);
+  CopyMetadataTo(call_inst, val);
```
[the remark this answers is not in the document]
Comment (kumarak) [reply libraries/anvill_passes/src/LowerRemillMemoryAccessIntrinsics.cpp:77]: This copies the metadata to store instruction. I updated it to assign to a variable and use that to copy the metadata.

File: libraries/anvill_passes/src/InstructionFolderPass.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pgoodman) [reply libraries/anvill_passes/src/InstructionFolderPass.cpp:556]: Are there any vector select cases that we actually handle?

**Review by pgoodman — APPROVED** [review]

**Comment by pgoodman** [comment]
@kumarak can you enable the now-passing tests in this PR, and look into the failing test?

```
[!] There were [1] failures on x86:
`Instructions.cpp:2873`: `1` failures
    /__w/anvill/anvill/src/anvill/angha-test-1k/results/x86/python/success/freebsd/usr.bin/systat/extr_iostat.c_devstats.elf/output.json
Ignored 3 tests.
    2 ignored tests are passing! They are:
    /__w/anvill/anvill/src/anvill/angha-test-1k/results/x86/python/success/postgres/src/backend/access/nbtree/extr_nbtsort.c__bt_parallel_heapscan.elf/output.json
    /__w/anvill/anvill/src/anvill/angha-test-1k/results/x86/python/success/sumatrapdf/ext/libwebp/src/enc/extr_picture_psnr_enc.c_AccumulateLSIM.elf/output.json
```

**Review by pgoodman — APPROVED** [review]

**Pull request merged**